### PR TITLE
Tests for PendingRequestQueue

### PR DESCRIPTION
**Pull request by sburmanoctopus**
[sc-53457]

# Background

In preparation for async halibut, we wanted to increase the test coverage of the pending request queue.

# Results

## Before
There was surprisingly low test coverage.

## After
Writing all tests for the pending request queue highlighted in [this document](https://github.com/OctopusDeploy/Halibut/pull/299/files).

We had to add the ability to override the timeout for dequeueing. Otherwise, this is just test writing.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs [hunk 1]
```diff
@@ -12,117 +16,408 @@ namespace Halibut.Tests.ServiceModel
...
+                var request = await sut.DequeueAsync();
+                request.Should().Be(expectedRequest);
             }
 
+            await ApplyResponsesToCompleteWaitingTasks(sut, requestsInOrder, queueAndWaitTasks);
+        }
+
+        [Test]
+        [TestCaseSource(typeof(AsyncTestCase))]
+        public async Task QueueAndWait_AddingMultipleItemsToQueueConcurrently_AllRequestsShouldBeSuccessfullyQueued(bool async)
```
Comment (LukeButters) [inline source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs:199]: Would be worth testing concurrently applying responses.

Does this test that the responses go to the correct requests?

File: source/Halibut/ServiceModel/PendingRequestQueue.cs [hunk 2]
```diff
@@ -68,6 +75,17 @@ public bool IsEmpty
             }
         }
 
+        public int Count
+        {
+            get
+            {
+                lock (sync)
```
Comment (LukeButters) [inline source/Halibut/ServiceModel/PendingRequestQueue.cs:82]: Why was this required?

Comment (LukeButters) [reply source/Halibut/ServiceModel/PendingRequestQueue.cs:82]: oh similar to is empty.

**Review by LukeButters — APPROVED** [review]

File: source/Halibut/ServiceModel/PendingRequestQueue.cs (hunk 2 above)
Comment (sburmanoctopus) [reply source/Halibut/ServiceModel/PendingRequestQueue.cs:82]: Yeah, I needed to know in the tests when the queue had filled up completely before we could start dequeueing.

File: source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs (hunk 1 above)
Comment (sburmanoctopus) [reply source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs:199]: > Would be worth testing concurrently applying responses.

Yeah, it would be done!

> Does this test that the responses go to the correct requests?

Actually, I don't think it does. Done!


I was a bit cheeky, and covered both of these requirements in `ApplyResponsesConcurrentlyAndEnsureAllQueueResponsesMatch`. Feel free to change it/split it out into separate test cases if you'd like 🙂

**Review by LukeButters — APPROVED** [review]
